### PR TITLE
[opt] [refactor] Move advanced optimization toggle to "ti.cfg.advanced_optimization"

### PR DESCRIPTION
**Pull request by archibate**
Related issue = #

[[Click here for the format server]](http://kun.csail.mit.edu:31415/)

----

### CODE REVIEW

File: taichi/ir/ir.cpp [hunk 1]
```diff
@@ -20,6 +20,10 @@ IRBuilder &current_ast_builder() {
   return context->builder();
 }
 
+CompileConfig &IRNode::get_config() const {
+  return get_kernel()->program.config;
+}
```
Comment (TH3CHARLie) [inline taichi/ir/ir.cpp:25]: Nice to have this! And BTW I like your proposal in #455 about a per-kernel compile config, I believe that's how we eventually entirely obsolete ti.cfg. This helper would be helpful when supporting that as well.

File: misc/constant_fold.py [hunk 2]
```diff
@@ -0,0 +1,19 @@
+# https://github.com/taichi-dev/taichi/pull/839#issuecomment-626217806
+import taichi as ti
+ti.init(print_ir=True)
+#ti.core.toggle_advanced_optimization(False)
```
Comment (TH3CHARLie) [inline misc/constant_fold.py:4]: Is this file some kind of demo of turning on/off adavanced_opt? Like the `listgen_demo.py`

Comment (archibate) [reply misc/constant_fold.py:4]: Yes, it shows off the power of JIT constant folding pass.

File: taichi/ir/ir.cpp (hunk 1 above)
Comment (archibate) [reply taichi/ir/ir.cpp:25]: Yes, I believe we may finally use `get_kernel()->config` for this.

File: misc/constant_fold.py (hunk 2 above)
Comment (TH3CHARLie) [reply misc/constant_fold.py:4]: > Yes, it shows off the power of JIT constant folding pass.

Then probably uses `demo` as the suffix as well? Actually, to keep consistency with the existing `test_*`, `benchmark_*` files, maybe name them as `demo_*` can be more clear.

Comment (xumingkuan) [reply misc/constant_fold.py:4]: > Then probably uses `demo` as the suffix as well?

Sounds good!

File: taichi/ir/ir.cpp (hunk 1 above)
Comment (yuanming-hu) [reply taichi/ir/ir.cpp:25]: Agreed!

**Comment by xumingkuan** [comment]
> Maybe use `optimization_level` instead, like `gcc`?

SGTM.

**Comment by archibate** [comment]
Let's do the `optimization_level` in #1396, i think `opt_level=3` or `optimizaion=3` is good enough? e.g.:
`opt_level=3` for fast_math + advanced_optimization.
`opt_level=2` for advanced_optimization.
`opt_level=1` for only basic optimization.
`opt_level=0` for disable all optimization.

**Comment by yuanming-hu** [comment]
Sounds good. One concern:

> `opt_level=3` for fast_math + advanced_optimization.

`fastmath` is independent from optimization, since it changes programs' behavior yet optimization doesn't.

**Comment by archibate** [comment]
Oh! I didn't know, but I found fast_math actually provide 14fps -> 15fps performance boost in mpm128?

**Comment by xumingkuan** [comment]
> Oh! I didn't know, but I found fast_math actually provide 14fps -> 15fps performance boost in mpm128?

Of course, fast_math provides optimization. But if a program involves `inf` or `NaN`, or even simply `1e50 - 1e50 + 1` (which should be `1` but under fast_math it can be `0`), the result can be wrong.

I found this paragraph at https://www.osc.edu/resources/getting_started/howto/howto_tune_performance:
> Faster operations are sometimes less accurate. For Intel compilers, fast math is default with -O2 and -O3. If you have a problem, use -fp-model precise. For GNU compilers, precise math is default with -O2 and -O3. If you want faster performance, use -ffast-math.

It makes me think of our `rsqrt`. Not sure how to make use of it for optimization though...

**Pull request merged**